### PR TITLE
RDoc-2105-Add-Link-Querying-Distinct also clarified and updated forma…

### PR DESCRIPTION
**Pull request by serezhub**
…tting

### CODE REVIEW

**Review by grisha-kotler — APPROVED** [review]

**Review by reebhub — CHANGES_REQUESTED** [review]

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 1]
```diff
@@ -24,14 +41,26 @@ RavenDB supports returning counts when the distinct operation is used.
...
 
-{INFO:Performance}
+{INFO: }
+
+### Performance Cost and an Alternative Approach
 
 Please keep in mind that this operation might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
 
-The same result might be achieved by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value of. e.g.
+The same result might be achieved by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:50]: map-reduce is only relevant for Distinct().Count().  
I also added a link to Oren's article on the topic, which Grisha approved.

I clarified that map-reduce is an alternative to distinct.count by changing the intro to this paragraph to this: 
> You can also ensure that queries do not count duplicate items by creating a [Map-Reduce]

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.java.markdown [hunk 2]
```diff
@@ -10,24 +25,38 @@ select distinct ShipTo.Country
...
-RavenDB supports returning counts when the distinct operation is used.
+To ensure that queries do not count duplicate items, 
+RavenDB supports using the `Distinct()` operation in combination with the `Count()` operation.
 
 {CODE:java distinct_2_1@Indexes\Querying\Distinct.java /}
 
-{INFO:Performance}
+{INFO: }
+
+### Performance Cost and an Alternative Approach
```
Comment (reebhub) [inline Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.java.markdown:43]: * please remove the "info" clause
* the title can be ## or you can leave it ### and add a --- yourself

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 3]
```diff
@@ -11,27 +26,42 @@ select distinct ShipTo.Country
...
 
-Please keep in mind that this operation might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
+You can also ensure that queries do not count duplicate items by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value. 
+[Indexes](../../indexes/creating-and-deploying) need to process entire datasets just once, after which they only process any new data. 
+Queries process all of the data assigned to them each time they are activated.
 
-The same result might be achieved by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value of. e.g.
+Learn how to use the alternative approach efficiently in the article [Implementing a count(distinct) query in RavenDB](https://ravendb.net/articles/implementing-a-countdistinct-query-in-ravendb).
+
+#### Map-Reduce Index Sample:
```
Comment (reebhub) [inline Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:57]: this is currently under the "Performance Cost and an Alternative Approach" section, does it belong here?

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.java.markdown (hunk 2 above)
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.java.markdown:43]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown (hunk 3 above)
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:57]: Yes.  I asked Oren if the distinct.count alternative is about map-reduce indexing exclusively.  He said that map-reduce is **the** alternative to distinct.count, so the map-reduce example indeed belongs in this section.

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 4]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
 
-Please keep in mind that this operation might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
+Learn how to use the alternative approach efficiently in the article [Implementing a count(distinct) query in RavenDB](https://ravendb.net/articles/implementing-a-countdistinct-query-in-ravendb).
 
-The same result might be achieved by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value of. e.g.
+#### Map-Reduce Index Sample:
+
+Index definition:
 
 {CODE:csharp distinct_3_1@Indexes\Querying\Distinct.cs /}
```
Comment (Danielle9897) [inline Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:59]: Code example should be similar to:
[https://demo.ravendb.net/demos/csharp/static-indexes/map-reduce-index](https://demo.ravendb.net/demos/csharp/static-indexes/map-reduce-index)

The `IndexEntry` class should have both the Country & the CountryCount
This way the aggregated count value is calculated by the index and is always up-to-date,
which is the point here...

Also, modify the query part accordingly.

**Review by Danielle9897 — CHANGES_REQUESTED** [review]

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 5]
```diff
@@ -1,7 +1,22 @@
 # Querying: Distinct
+---
+
+{NOTE: }
 
 The `Distinct` method allows you to remove duplicates from the result. Items are compared based on the fields listed in the `select` section of the query.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:6]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 6]
```diff
@@ -1,7 +1,22 @@
 # Querying: Distinct
+---
+
+{NOTE: }
 
 The `Distinct` method allows you to remove duplicates from the result. Items are compared based on the fields listed in the `select` section of the query. 
 
+* In this page:
+   * [Sample Query with Distinct Method](../../indexes/querying/distinct#sample-query-with-the-distinct-method)
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:9]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 7]
```diff
@@ -1,7 +1,22 @@
 # Querying: Distinct
+---
+
+{NOTE: }
 
 The `Distinct` method allows you to remove duplicates from the result. Items are compared based on the fields listed in the `select` section of the query. 
 
+* In this page:
+   * [Sample Query with Distinct Method](../../indexes/querying/distinct#sample-query-with-the-distinct-method)
+   * [Paging with the Distinct Method](../../indexes/querying/distinct#paging-with-the-distinct-method)
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:10]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 8]
```diff
@@ -1,7 +1,22 @@
...
+---
+
+{NOTE: }
 
 The `Distinct` method allows you to remove duplicates from the result. Items are compared based on the fields listed in the `select` section of the query. 
 
+* In this page:
+   * [Sample Query with Distinct Method](../../indexes/querying/distinct#sample-query-with-the-distinct-method)
+   * [Paging with the Distinct Method](../../indexes/querying/distinct#paging-with-the-distinct-method)
+   * [Count](../../indexes/querying/distinct#count)
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:11]: done.

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 9]
```diff
@@ -1,7 +1,22 @@
...
+   * [Performance Cost and an Alternative Approach](../../indexes/querying/distinct#performance-cost-and-an-alternative-approach)
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:12]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 10]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
 {CODE-TAB-BLOCK/}
 {CODE-TABS/}
 
-## Paging 
+## Paging with the Distinct Method
+
+A special approach must be used when calling the `distinct()` method while paging.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:31]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 11]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
+A special approach must be used when calling the `distinct()` method while paging.  
+Please read the dedicated article about [paging through tampered results](../../indexes/querying/paging#paging-through-tampered-results).  
 
-Please read the dedicated article about [paging through tampered results](../../indexes/querying/paging#paging-through-tampered-results). This kind of paging is required when using a distinct keyword.
 
 ## Count
 
-RavenDB supports returning counts when the distinct operation is used.
+To ensure that queries do not count duplicate items, 
+RavenDB supports using the `Distinct()` operation in combination with the `Count()` operation.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:38]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 12]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
 
 {CODE-TABS}
 {CODE-TAB:csharp:Query distinct_2_1@Indexes\Querying\Distinct.cs /}
 {CODE-TAB:csharp:DocumentQuery distinct_2_2@Indexes\Querying\Distinct.cs /}
 {CODE-TABS/}
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:43]: 1. In RQL I only got as far as `from "Orders" select distinct ShipTo.Country`.  As soon as I added the count word, I got an error.  
2. done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 13]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
 
-{INFO:Performance}
+## Performance Cost and an Alternative Approach
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:45]: We don't have an article about Count
It is used in the [projection](https://ravendb.net/docs/article-page/5.4/Csharp/indexes/querying/projections#example-vii---projection-with-a-count-predicate) example, it's in a while() condition in 3 [Paging](https://ravendb.net/docs/article-page/5.4/Csharp/indexes/querying/paging#paging-through-tampered-results) examples, and in the [filtering](https://ravendb.net/docs/article-page/5.4/Csharp/indexes/querying/filtering#where---nested-property) article.

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 14]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
+
+Please keep in mind that `Distinct()` might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
+
+You can also ensure that queries do not count duplicate items by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value. 
+[Indexes](../../indexes/creating-and-deploying) need to process entire datasets just once, after which they only process any new data. 
+Queries process all of the data assigned to them each time they are activated.
 
-Please keep in mind that this operation might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
+Learn how to use the alternative approach efficiently in the article [Implementing a count(distinct) query in RavenDB](https://ravendb.net/articles/implementing-a-countdistinct-query-in-ravendb).
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:53]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 15]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
 
 {CODE-TABS}
 {CODE-TAB:csharp:Query distinct_2_1@Indexes\Querying\Distinct.cs /}
 {CODE-TAB:csharp:DocumentQuery distinct_2_2@Indexes\Querying\Distinct.cs /}
 {CODE-TABS/}
 
-{INFO:Performance}
+## Performance Cost and an Alternative Approach
+
+Please keep in mind that `Distinct()` might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:47]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 16]
```diff
@@ -11,36 +26,58 @@ select distinct ShipTo.Country
...
+
+You can also ensure that queries do not count duplicate items by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index that aggregates data by the field where you want a distinct value. 
+[Indexes](../../indexes/creating-and-deploying) need to process entire datasets just once, after which they only process any new data. 
+Queries process all of the data assigned to them each time they are activated.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:51]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 17]
```diff
@@ -11,36 +27,67 @@ select distinct ShipTo.Country
...
 {CODE-TAB:csharp:DocumentQuery distinct_3_3@Indexes\Querying\Distinct.cs /}
 {CODE-TABS/}
 
-{INFO/}
+---
+
+### Combining Faceted Queries with Map-Reduce
+
+Faceted queries can be used together with a map-reduce index as another alternative approach.  
+See [Implementing a count(distinct) query in RavenDB](https://ravendb.net/articles/implementing-a-countdistinct-query-in-ravendb).
```
Comment (Danielle9897) [inline Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:75]: =>
See article [xxx]() for such an example.

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 18]
```diff
@@ -11,36 +27,67 @@ select distinct ShipTo.Country
...
+
+Please keep in mind that using `Count()` with `Distinct()` might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
 
-Please keep in mind that this operation might not be efficient for large sets of data due to the need to scan all of the index results in order to find all the unique values.
+* Getting the distinct items' count can also be achieved by creating a [Map-Reduce](../../indexes/map-reduce-indexes) index 
+  that will aggregate data by the field for which distinct count results are needed.
+* This is more efficient since computations are done during indexing time and not at query time.  
+  The entire dataset is [Indexed](../../indexes/creating-and-deploying) 
+  once, whereafter the aggregated value is always kept up to date as indexing will occur only for new/modified data.  
+  Queries, on the other hand, process all of the data assigned to them each time they are activated.
```
Comment (Danielle9897) [inline Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:55]: Remove the above sentence about "Queries, on the other hand...."

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs [hunk 19]
```diff
@@ -91,19 +100,22 @@ public Distinct()
...
                 }
 
                 using (var session = store.OpenSession())
                 {
                     #region distinct_3_3
-                    var numberOfCountries = session
-                        .Advanced
-                        .DocumentQuery<Order_Countries.Result, Order_Countries>()
-                        .Count();
+                    // The query calls the map-reduce index defined above
```
Comment (Danielle9897) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs:114]: =>
Query the map-reduce index defined above

**Review by Danielle9897 — CHANGES_REQUESTED** [review]

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown (hunk 4 above)
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:61]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 20]
```diff
@@ -1,6 +1,22 @@
 # Querying: Distinct
+---
 
-The `Distinct` method allows you to remove duplicates from the result. Items are compared based on the fields listed in the `select` section of the query. 
+{NOTE: }
+
+The `Distinct` method allows you to remove duplicates from the result.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:6]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 21]
```diff
@@ -11,36 +27,67 @@ select distinct ShipTo.Country
...
+A special approach must be used when calling `distinct()` while paging.  
+Please read the dedicated article about [paging through tampered results](../../indexes/querying/paging#paging-through-tampered-results).  
 
-## Count
 
-RavenDB supports returning counts when the distinct operation is used.
+## Count with Distinct
+
+Use `Count()` in combination with `Distinct()` to get the number of unique items.  
+Similar to ToList(), Count() executes the query on the server-side.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:39]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown [hunk 22]
```diff
@@ -11,36 +27,67 @@ select distinct ShipTo.Country
...
 
 {CODE-TABS}
 {CODE-TAB:csharp:Query distinct_2_1@Indexes\Querying\Distinct.cs /}
 {CODE-TAB:csharp:DocumentQuery distinct_2_2@Indexes\Querying\Distinct.cs /}
 {CODE-TABS/}
 
-{INFO:Performance}
+## Performance Cost And Alternative Approaches
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:46]: done

File: Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown (hunk 17 above)
Comment (serezhub) [reply Documentation/4.0/Raven.Documentation.Pages/indexes/querying/distinct.dotnet.markdown:75]: done

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs (hunk 19 above)
Comment (serezhub) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs:114]: done

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs [hunk 23]
```diff
@@ -91,19 +100,22 @@ public Distinct()
                 using (var session = store.OpenSession())
                 {
                     #region distinct_3_2
-                    var numberOfCountries = session
-                        .Query<Order_Countries.Result, Order_Countries>()
-                        .Count();
+                    // The query calls the map-reduce index defined above
```
Comment (Danielle9897) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs:103]: =>
Query the map-reduce index defined above

**Review by Danielle9897 — CHANGES_REQUESTED** [review]

**Review by Danielle9897 — APPROVED** [review]

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs (hunk 23 above)
Comment (serezhub) [reply Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Indexes/Querying/Distinct.cs]: done

**Pull request merged**